### PR TITLE
ceph: fix panic when recreating the csidriver object

### PR DESCRIPTION
**Pull request by Madhu-1**
**Description of your changes:**

This PR fixes the below items

* fix panic in reCreateCSIDriverInfo
   Initialize the client and the csidriver object before calling the reCreateCSIDriverInfo function.
* Reuse defined csiDriver object
    reuse already defined csiDriver object instead of referring to the method receiver
* fix logging of csidriver
    log the successful message about starting CSIDriver after creating the daemonset and deployment objects.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Note: the issue was reported at slack https://rook-io.slack.com/archives/CK9CF5H2R/p1629776387282600

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Comment by Madhu-1** [comment]
Note:- This needs to be backported to 1.6 and 1.7

File: pkg/operator/ceph/csi/betav1csidriver.go [hunk 1]
```diff
@@ -72,6 +72,8 @@ func (d beta1CsiDriver) createCSIDriverInfo(ctx context.Context, clientset kuber
 	// As FSGroupPolicy field is immutable, should be set only during create time.
 	// if the request is to change the FSGroupPolicy, we are deleting the CSIDriver object and creating it.
 	if driver.Spec.FSGroupPolicy != nil && csiDriver.Spec.FSGroupPolicy != nil && *driver.Spec.FSGroupPolicy != *csiDriver.Spec.FSGroupPolicy {
+		d.csiClient = csidrivers
```
Comment (Yuggupta27) [inline pkg/operator/ceph/csi/betav1csidriver.go:75]: If we are initializing the client and csidriver objects before the call, do we need the initialization in the reCreateCSIDriverInfo func [here](https://github.com/rook/rook/blob/master/pkg/operator/ceph/csi/betav1csidriver.go#L91-L92)?

File: pkg/operator/ceph/csi/spec.go [hunk 2]
```diff
@@ -516,6 +514,7 @@ func startDrivers(clientset kubernetes.Interface, rookclientset rookclient.Inter
 			return errors.Wrapf(err, "failed to start rbd provisioner deployment: %+v", rbdProvisionerDeployment)
 		}
 		k8sutil.AddRookVersionLabelToDeployment(rbdProvisionerDeployment)
+		logger.Info("successfully started CSI Ceph RBD")
```
Comment (Yuggupta27) [inline pkg/operator/ceph/csi/spec.go:517]: "successfully started CSI Ceph RBD" -> "successfully started CSI Ceph RBD driver"

File: pkg/operator/ceph/csi/betav1csidriver.go (hunk 1 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/betav1csidriver.go:75]: we are just creating the variables in the function you defined above if we don't initialize the client and driver object we will hit the nil pointer issue.

Comment (Yuggupta27) [reply pkg/operator/ceph/csi/betav1csidriver.go:75]: So the variable declaration in the reCreateCSIDriverInfo function is just for using shortcuts like `csiDriver.Name` in place of `d.csiDriver.Name`. Declaring new variables in the function looks confusing at first, as it looks like the same variable is being declared 2 times(which is not actually happening). It looks cleaner with the variables :+1:

File: pkg/operator/ceph/csi/spec.go [hunk 3]
```diff
@@ -584,6 +583,7 @@ func startDrivers(clientset kubernetes.Interface, rookclientset rookclient.Inter
 			return errors.Wrapf(err, "failed to start cephfs provisioner deployment: %+v", cephfsProvisionerDeployment)
 		}
 		k8sutil.AddRookVersionLabelToDeployment(cephfsProvisionerDeployment)
+		logger.Info("successfully started CSI CephFS driver")
```
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go:586]: Shall we have a separate message for the volume plugin (line 555) and the provisioner here?
```suggestion
		logger.Info("successfully started CSI CephFS provisioner")
```

File: pkg/operator/ceph/csi/spec.go [hunk 4]
```diff
@@ -516,6 +514,7 @@ func startDrivers(clientset kubernetes.Interface, rookclientset rookclient.Inter
 			return errors.Wrapf(err, "failed to start rbd provisioner deployment: %+v", rbdProvisionerDeployment)
 		}
 		k8sutil.AddRookVersionLabelToDeployment(rbdProvisionerDeployment)
+		logger.Info("successfully started CSI Ceph RBD driver")
```
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go:517]: Same suggestion as below
```suggestion
		logger.Info("successfully started CSI Ceph RBD provisioner")
```

**Review by Yuggupta27 — APPROVED** [review]

File: pkg/operator/ceph/csi/csidriver.go [hunk 5]
```diff
@@ -95,7 +97,7 @@ func (d v1CsiDriver) reCreateCSIDriverInfo(ctx context.Context) error {
 		return errors.Wrapf(err, "failed to delete CSIDriver object for driver %q", csiDriver.Name)
 	}
 	logger.Infof("CSIDriver object deleted for driver %q", csiDriver.Name)
-	_, err = csiClient.Create(ctx, d.csiDriver, metav1.CreateOptions{})
```
Comment (travisn) [inline pkg/operator/ceph/csi/csidriver.go:98]: How about leaving this `d.csiDriver` reference and delete the two local vars `csiDriver` and `csiClient`? The local vars are not necessary IMO.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/spec.go (hunk 3 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go:586]: @travisn as we are starting both provisioner and daemonset I left the logging as it was previously

**Review by travisn — APPROVED** [review]
Were you able to repro before the fix, or this is based on the stack trace?

**Pull request merged**